WT-5674 Prepare support to add functional tests

raviprakashgiri29 opened the pull request.

ddanderson commented on test/suite/test_prepare_hs03.py:
```diff
@@ -0,0 +1,183 @@
+        cursor = self.session.open_cursor(self.uri)
+        for i in range(1, nkeys):
+            self.session.begin_transaction('isolation=snapshot')
+            tkey = 'C' + ds.key(nrows + i)
```
The key here has a `'C'` prefix, which means it won't match the original keys.  So these changes are not updates, and they won't create "old" versions of values, and thus won't generate history store activity.  Is there some other reasoning for using a new key?

raviprakashgiri29 replied:
@ddanderson : Removed the prefix and changed the code to update the same key with different values for history store activity.

ddanderson commented on test/suite/test_prepare_hs03.py:
```diff
@@ -0,0 +1,183 @@
+            start = (j * nkeys)
+            end = start + nkeys
+            for i in range(start, end):
+                tkey = 'P' + ds.key(nrows + i)
```
Similar to comment above, these keys won't match the original keys.

raviprakashgiri29 replied:
Fixed

ddanderson commented on test/suite/test_prepare_hs03.py:
```diff
@@ -0,0 +1,183 @@
+            tkey = 'P' + ds.key(nrows + i)
+            cursor.set_key(tkey)
+            # The search should fail i.e, WT_NOTFOUND
+            self.assertEqual(cursor.search(), wiredtiger.WT_NOTFOUND)
```
If you end up deleting the prefixes for the keys, the search here will succeed, and the correctness test will need to be whether the committed value is visible, rather than the prepared value.  If you go this route, you won't need to do a final scan searching for keys inserted with the commit (the loop at line 129), since you they will be all the same keys.  Although the scan after the simulated crash would be still be good.

raviprakashgiri29 replied:
Updated the code and added some correctness tests to check that committed value is visible and prepared value not visible.

raviprakashgiri29 commented on test/suite/test_prepare_hs03.py:
```diff
@@ -0,0 +1,183 @@
+        filename="test_prepare_hs03.wt"
+        self.assertEquals(os.path.exists(filename), True)
+
+        with open(filename, 'r+') as log:
```
Fixed

raviprakashgiri29 commented:
> Hi @raviprakashgiri29, this is on the right track, but one important part is to use the same set of keys for commits and prepares. That truly will create updates to records, not new records. It is updates that will create history store traffic. We certainly want to test prepares on new records as well and make sure salvage and verify work there.
> 
> Another thing I'd like to see is salvage and verify tests _without_ corrupting any files as a real corruption should be rare. Perhaps this can be done with scenarios - one scenario corrupting and one not.

@ddanderson Yeah, I agree with your concern here. I have updated the code to use the same key, update it with different values and added couple of correction tests. Also, added scenarios to run test with corrupting the table and without-corrupting the table.

raviprakashgiri29 commented on test/suite/test_prepare_hs03.py:
```diff
@@ -0,0 +1,196 @@
+        # Insert some records with commit timestamp, corrupt file and call salvage, verify before checkpoint.
+
+        # Commit some updates to get eviction and history store fired up
+        commit_value = b"bbbbb" * 2
```
@ddanderson Thanks for pointing out to use `stat.conn.cache_write_hs`. I updated the code to use the history store statistics, updated `commit_value` & `prepare_value` to use the numbers from `test_prepare_hs01.py`.

raviprakashgiri29 commented on test/suite/test_prepare_hs03.py:
```diff
@@ -0,0 +1,196 @@
+        # prepared updates to the history store
+        sessions = [0] * nsessions
+        cursors = [0] * nsessions
+        prepare_value = b"ccccc" * 2
```
Fixed

raviprakashgiri29 commented on test/suite/test_prepare_hs03.py:
```diff
@@ -0,0 +1,196 @@
+        # We put prepared updates in multiple sessions so that we do not hang
+        # because of cache being full with uncommitted updates.
+        nsessions = 3
+        nkeys = 4
```
Fixed

raviprakashgiri29 commented:
> @raviprakashgiri29, great improvements. I think we need to make sure we get enough history traffic to really drive this (see review comments). Since we haven't gotten prepare with HS fully working yet, when you get the sizing big enough, it will probably hang! When you get it to that point, we'll know it's a good test (at least it's generating history store traffic). To pass the pull request testing, you'll need to insert a `@unittest.skip(...)`, we'll know that enabling/passing this test will be a condition of getting the prepare/hs work completed.
> 
> See also test_prepare_hs01.py, which hangs when you remove the `@unittest.skip("Temporarily disabled")` line, for the same reason.
> 
> If you like, maybe we can have a quick chat about this?

@ddanderson I have updated the code and added more traffic to write content to the history store, used statistics to confirm history store writes and skipped this test to pass. 

Sure, we can discuss anytime tomorrow as per your availability.

ddanderson commented on test/suite/test_prepare_hs03.py:
```diff
@@ -0,0 +1,211 @@
+            # Correctness Test - commit_value should be visible
+            self.assertEquals(cursor.get_value(), commit_value)
+            # Correctness Test - prepare_value should NOT be visible
+            self.assertNotEquals(cursor.get_value(), prepare_value)
```
This is redundant code, there's no way the previous assert can pass and this one will fail.

raviprakashgiri29 replied:
I do agree with your concern here, but this code is to confirm that prepare_value is not visible after prepare_transactions and before commit.

ddanderson commented on test/suite/test_prepare_hs03.py:
```diff
@@ -0,0 +1,211 @@
+            # Correctness Test - commit_value should be visible
+            self.assertEquals(cursor.get_value(), commit_value)
+            # Correctness Test - prepare_value should NOT be visible
+            self.assertNotEquals(cursor.get_value(), prepare_value)
```
Another redundant statement that should be removed.

ddanderson commented:
@raviprakashgiri29, I just reviewed this.  I'm assuming the test hangs if you enable it?
After you finish the review and merge, can you please put a comment in WT-5673 (salvage/verify work for prepare) that this test must be enabled?

raviprakashgiri29 commented:
> @raviprakashgiri29, I just reviewed this. I'm assuming the test hangs if you enable it?
> After you finish the review and merge, can you please put a comment in WT-5673 (salvage/verify work for prepare) that this test must be enabled?

@ddanderson Thank you for reviewing the changes. Yes, the test hangs if I enable it locally.
I will update WT-5673 to enable this test.

ddanderson commented:
@raviprakashgiri29, this looks good.  I think you can merge it now.

kommiharibabu commented on test/suite/test_prepare_hs03.py:
```diff
@@ -0,0 +1,209 @@
+        cursor = self.session.open_cursor(self.uri)
+
+        # Search the keys inserted with commit timestamp after crash
+        self.session.begin_transaction('read_timestamp=' + timestamp_str(1))
```
All the read timestamps must point to the latest timestamp. It is to make sure that even with the latest timestamp, we could only see the commit value only.

raviprakashgiri29 replied:
Fixed
